flambe/nn/transformer_sru.py

```diff
@@ -0,0 +1,307 @@
+                                                   dim_feedforward,
+                                                   dropout)
+
+        self.layers = _get_clones(decoder_layer, num_layers)
+        self._reset_parameters()
+
+
+class TransformerSRUEncoderLayer(Module):
+    """A TransformerSRUEncoderLayer with an SRU replacing the FFN."""
+
+    def __init__(self,
+                 d_model: int,
+                 nhead: int,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 bidirectional: bool = False,
+                 **kwargs: Dict[str, Any]) -> None:
+        """Initialize a TransformerSRUEncoderLayer.
+
+        Parameters
+        ----------
+        d_model : int
+            The number of expected features in the input.
+        n_head : int
+            The number of heads in the multiheadattention models.
+        dim_feedforward : int, optional
+            The dimension of the feedforward network (default=2048).
+        dropout : float, optional
+            The dropout value (default=0.1).
+        bidirectional: bool
+            Whether the SRU module should be bidrectional.
+            Defaul ``False``.
+
+        Extra keyword arguments are passed to the SRUCell.
+
+        """
+        super(TransformerSRUEncoderLayer, self).__init__()
+
+        self.self_attn = nn.MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.sru = SRUCell(d_model,
+                           dim_feedforward,
+                           dropout,
+                           bidirectional=bidirectional,
+                           has_skip_term=False, **kwargs)
+
+        self.linear2 = nn.Linear(dim_feedforward, d_model)
+        self.norm1 = nn.LayerNorm(d_model)
+        self.norm2 = nn.LayerNorm(d_model)
+        self.dropout1 = nn.Dropout(dropout)
+        self.dropout2 = nn.Dropout(dropout)
+
+    def forward(self,  # type: ignore
+                src: torch.Tensor,
+                memory: Optional[torch.Tensor] = None,
+                src_mask: Optional[torch.Tensor] = None,
+                src_key_padding_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+        """Pass the input through the endocder layer.
+
+        Parameters
+        ----------
+        src: torch.Tensor
+            The seqeunce to the encoder layer (required).
+        memory: Optional[torch.Tensor]
+            Optional memory from previous sequence.
+        src_mask: torch.Tensor, optional
+            The mask for the src sequence (optional).
+        src_key_padding_mask: torch.Tensor, optional
+            The mask for the src keys per batch (optional).
+
+        """
+        src2 = self.self_attn(src, src, src, attn_mask=src_mask,
+                              key_padding_mask=src_key_padding_mask)[0]
+        src = src + self.dropout1(src2)
+        src = self.norm1(src)
+        src2 = self.linear2(self.sru(src, state=memory, mask_pad=src_key_padding_mask))
```

Comment:
state is not an argument for SRUcell. (should have renamed c0 into state in the future for SRU)
for now you can do self.sru(tgt, memory, mask_pad=..) instead